bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

```diff
@@ -244,6 +244,7 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 	// Attempt to Patch the KubeadmConfig object and status after each reconciliation if no error occurs.
 	defer func() {
 		if rerr == nil {
+			config.Status.ObservedGeneration = config.Generation
```

Comment:
IMO this ObservedGeneration, should be always set, no matter if the reconciliation succeded or not (it should record the latest generation observed by the controller)
TBH, I find a little bit inconsistent that for this controller we are patching only if there is no error

Reply:
This is where it is tricky. `ObservedGeneration` is supposed to be the latest version that has been reconciled. If there was an error and we are going to requeue reconciliation, it means we weren't able to successfully reconcile the object and are going to try again. I think the right semantic of this is then to treat it as though we haven't successfully applied that version of the object.

Reply:
It seems this isn't quite consistent. [DaemonSets only update it when "expectations" are met],(https://github.com/kubernetes/kubernetes/blob/4bc907f3c956ffb808fdcfb4f8fc07825b8d2339/pkg/controller/daemon/daemon_controller.go#L1157-L1160), not sure exactly how that works though, but it's clear that sometimes it is not updated.

StatefulSets only update it [if the call to `updateStatefulSet` succeeds without erroring](https://github.com/kubernetes/kubernetes/blob/4bc907f3c956ffb808fdcfb4f8fc07825b8d2339/pkg/controller/statefulset/stateful_set_control.go#L103-L106), which you can see there's a lot of reasons it might error. It sets the ObservedGeneration, but doesn't persist the changes if an error occurs.

GCP's Cloud Run explicitly calls out that it updates [even if the controller fails to process the object](https://cloud.google.com/run/docs/reference/rest/v1/namespaces.configurations#ConfigurationStatus.FIELDS.observed_generation).